actors/market/src/balance_table.rs

```diff
@@ -0,0 +1,173 @@
+    pub fn root(&mut self) -> Result<Cid, HamtError> {
+        self.0.flush()
+    }
+
+    /// Gets token amount for given address in balance table
+    pub fn get(&self, key: &Address) -> Result<TokenAmount, HamtError> {
+        if let Some(v) = self.0.get(&key.to_bytes())? {
+            Ok(v.0.clone())
+        } else {
+            Ok(0.into())
+        }
+    }
+
+    /// Adds token amount to previously initialized account.
+    pub fn add(&mut self, key: &Address, value: &TokenAmount) -> Result<(), HamtError> {
+        let prev = self.get(key)?;
+        let sum = &prev + value;
+        if sum.is_negative() {
+            Err(format!("New balance in table cannot be negative: {}", sum).into())
+        } else if sum.is_zero() && !prev.is_zero() {
+            self.0.delete(&key.to_bytes())?;
+            Ok(())
+        } else {
+            self.0.set(key.to_bytes().into(), BigIntDe(sum))?;
+            Ok(())
+        }
+    }
+
+    /// Subtracts up to the specified amount from a balance, without reducing the balance
+    /// below some minimum.
+    /// Returns the amount subtracted (always positive or zero).
+    pub fn subtract_with_minimum(
+        &mut self,
+        key: &Address,
+        req: &TokenAmount,
+        floor: &TokenAmount,
+    ) -> Result<TokenAmount, HamtError> {
+        let prev = self.get(key)?;
+        let available = std::cmp::max(TokenAmount::zero(), prev - floor);
+        let sub: TokenAmount = std::cmp::min(&available, req).clone();
+
+        if sub.is_positive() {
+            self.add(key, &-sub.clone())?;
+        }
+
+        Ok(sub)
+    }
+
+    /// Subtracts value from a balance, and errors if full amount was not substracted.
+    pub fn must_subtract(&mut self, key: &Address, req: &TokenAmount) -> Result<(), HamtError> {
+        let prev = self.get(key)?;
+
+        if req > &prev {
+            Err("couldn't subtract the requested amount".into())
+        } else {
+            self.add(key, &-req)
+        }
+    }
+
+    /// Returns total balance held by this balance table
+    #[allow(dead_code)]
+    pub fn total(&self) -> Result<TokenAmount, HamtError> {
+        let mut total = TokenAmount::default();
+
+        self.0.for_each(|_, v: &BigIntDe| {
+            total += &v.0;
+            Ok(())
+        })?;
+
+        Ok(total)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use fvm_shared::address::Address;
+    use fvm_shared::blockstore::MemoryBlockstore;
+    use fvm_shared::econ::TokenAmount;
+
+    use crate::balance_table::BalanceTable;
+
+    #[test]
+    fn total() {
+        let addr1 = Address::new_id(100);
+        let addr2 = Address::new_id(101);
+        let store = MemoryBlockstore::default();
+        let mut bt = BalanceTable::new(&store);
+
+        assert_eq!(bt.total().unwrap(), TokenAmount::from(0u8));
+
+        struct TotalTestCase<'a> {
+            amount: u64,
+            addr: &'a Address,
+            total: u64,
+        }
+        let test_vectors = [
```

Comment:
nit: not your name but consider changing name to avoid confusion with protocol test vectors for new readers.